### ISSUE TITLE
make fails with test failure Error: Cannot find module 'callback-to-async-iterator'

### ISSUE THREAD

**Issue by prasannapad** [pulumi/pulumi#3464]
`make all` failed running tests

TEST ALL:
./node_modules/.bin/istanbul test --print none _mocha -- --timeout 15000 'bin/tests/**/*.spec.js'
Error: Cannot find module 'callback-to-async-iterator'
Require stack:
- /usr/local/src/pulumi/sdk/nodejs/bin/runtime/invoke.js
- /usr/local/src/pulumi/sdk/nodejs/bin/runtime/resource.js
- /usr/local/src/pulumi/sdk/nodejs/bin/resource.js
- /usr/local/src/pulumi/sdk/nodejs/bin/output.js
- /usr/local/src/pulumi/sdk/nodejs/bin/runtime/closure/createClosure.js
- /usr/local/src/pulumi/sdk/nodejs/bin/runtime/closure/serializeClosure.js
- /usr/local/src/pulumi/sdk/nodejs/bin/runtime/index.js
- /usr/local/src/pulumi/sdk/nodejs/bin/metadata.js
- /usr/local/src/pulumi/sdk/nodejs/bin/config.js
- /usr/local/src/pulumi/sdk/nodejs/bin/index.js
- /usr/local/src/pulumi/sdk/nodejs/bin/tests/config.spec.js
- /usr/local/src/pulumi/sdk/nodejs/node_modules/mocha/lib/mocha.js
- /usr/local/src/pulumi/sdk/nodejs/node_modules/mocha/index.js
- /usr/local/src/pulumi/sdk/nodejs/node_modules/mocha/bin/_mocha
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:794:15)
    at Function.Module._load (internal/modules/cjs/loader.js:687:27)
    at Module.require (internal/modules/cjs/loader.js:849:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.<anonymous> (/usr/local/src/pulumi/sdk/nodejs/bin/runtime/invoke.js:25:18)
    at Module._compile (internal/modules/cjs/loader.js:956:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:973:10)
    at Module.load (internal/modules/cjs/loader.js:812:32)
    at Function.Module._load (internal/modules/cjs/loader.js:724:14)
    at Module.require (internal/modules/cjs/loader.js:849:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.<anonymous> (/usr/local/src/pulumi/sdk/nodejs/bin/runtime/resource.js:32:18)
    at Module._compile (internal/modules/cjs/loader.js:956:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:973:10)
    at Module.load (internal/modules/cjs/loader.js:812:32)
    at Function.Module._load (internal/modules/cjs/loader.js:724:14)
    at Module.require (internal/modules/cjs/loader.js:849:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.<anonymous> (/usr/local/src/pulumi/sdk/nodejs/bin/resource.js:28:20)
    at Module._compile (internal/modules/cjs/loader.js:956:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:973:10)
    at Module.load (internal/modules/cjs/loader.js:812:32)
    at Function.Module._load (internal/modules/cjs/loader.js:724:14)
    at Module.require (internal/modules/cjs/loader.js:849:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.<anonymous> (/usr/local/src/pulumi/sdk/nodejs/bin/output.js:25:20)
    at Module._compile (internal/modules/cjs/loader.js:956:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:973:10)
    at Module.load (internal/modules/cjs/loader.js:812:32)
    at Function.Module._load (internal/modules/cjs/loader.js:724:14)
    at Module.require (internal/modules/cjs/loader.js:849:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.<anonymous> (/usr/local/src/pulumi/sdk/nodejs/bin/runtime/closure/createClosure.js:28:18)
    at Module._compile (internal/modules/cjs/loader.js:956:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:973:10)
    at Module.load (internal/modules/cjs/loader.js:812:32)
    at Function.Module._load (internal/modules/cjs/loader.js:724:14)
    at Module.require (internal/modules/cjs/loader.js:849:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.<anonymous> (/usr/local/src/pulumi/sdk/nodejs/bin/runtime/closure/serializeClosure.js:26:17)
    at Module._compile (internal/modules/cjs/loader.js:956:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:973:10)
    at Module.load (internal/modules/cjs/loader.js:812:32)
    at Function.Module._load (internal/modules/cjs/loader.js:724:14)
    at Module.require (internal/modules/cjs/loader.js:849:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.<anonymous> (/usr/local/src/pulumi/sdk/nodejs/bin/runtime/index.js:19:26)
    at Module._compile (internal/modules/cjs/loader.js:956:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:973:10)
    at Module.load (internal/modules/cjs/loader.js:812:32)
    at Function.Module._load (internal/modules/cjs/loader.js:724:14)
    at Module.require (internal/modules/cjs/loader.js:849:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.<anonymous> (/usr/local/src/pulumi/sdk/nodejs/bin/metadata.js:17:17)
    at Module._compile (internal/modules/cjs/loader.js:956:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:973:10)
    at Module.load (internal/modules/cjs/loader.js:812:32)
    at Function.Module._load (internal/modules/cjs/loader.js:724:14)
    at Module.require (internal/modules/cjs/loader.js:849:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.<anonymous> (/usr/local/src/pulumi/sdk/nodejs/bin/config.js:17:20)
    at Module._compile (internal/modules/cjs/loader.js:956:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:973:10)
    at Module.load (internal/modules/cjs/loader.js:812:32)
    at Function.Module._load (internal/modules/cjs/loader.js:724:14)
    at Module.require (internal/modules/cjs/loader.js:849:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.<anonymous> (/usr/local/src/pulumi/sdk/nodejs/bin/index.js:22:10)
    at Module._compile (internal/modules/cjs/loader.js:956:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:973:10)
    at Module.load (internal/modules/cjs/loader.js:812:32)
    at Function.Module._load (internal/modules/cjs/loader.js:724:14)
    at Module.require (internal/modules/cjs/loader.js:849:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.<anonymous> (/usr/local/src/pulumi/sdk/nodejs/bin/tests/config.spec.js:17:17)
    at Module._compile (internal/modules/cjs/loader.js:956:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:973:10)
    at Module.load (internal/modules/cjs/loader.js:812:32)
    at Function.Module._load (internal/modules/cjs/loader.js:724:14)
    at Module.require (internal/modules/cjs/loader.js:849:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at /usr/local/src/pulumi/sdk/nodejs/node_modules/mocha/lib/mocha.js:231:27
    at Array.forEach (<anonymous>)
    at Mocha.loadFiles (/usr/local/src/pulumi/sdk/nodejs/node_modules/mocha/lib/mocha.js:228:14)
    at Mocha.run (/usr/local/src/pulumi/sdk/nodejs/node_modules/mocha/lib/mocha.js:514:10)
    at Object.<anonymous> (/usr/local/src/pulumi/sdk/nodejs/node_modules/mocha/bin/_mocha:480:18)
    at Module._compile (internal/modules/cjs/loader.js:956:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:973:10)
    at Module.load (internal/modules/cjs/loader.js:812:32)
    at Function.Module._load (internal/modules/cjs/loader.js:724:14)
    at Function.Module.runMain (internal/modules/cjs/loader.js:1025:10)
    at runFn (/usr/local/src/pulumi/sdk/nodejs/node_modules/istanbul/lib/command/common/run-with-cover.js:122:16)
    at Object.run (/usr/local/src/pulumi/sdk/nodejs/node_modules/istanbul/lib/command/common/run-with-cover.js:254:9)
    at TestCommand.run (/usr/local/src/pulumi/sdk/nodejs/node_modules/istanbul/lib/command/test.js:27:22)
    at runCommand (/usr/local/src/pulumi/sdk/nodejs/node_modules/istanbul/lib/cli.js:83:19)
    at runToCompletion (/usr/local/src/pulumi/sdk/nodejs/node_modules/istanbul/lib/cli.js:87:5)
    at Object.<anonymous> (/usr/local/src/pulumi/sdk/nodejs/node_modules/istanbul/lib/cli.js:93:5)
    at Module._compile (internal/modules/cjs/loader.js:956:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:973:10)
    at Module.load (internal/modules/cjs/loader.js:812:32)
    at Function.Module._load (internal/modules/cjs/loader.js:724:14)
    at Function.Module.runMain (internal/modules/cjs/loader.js:1025:10)
    at internal/main/run_main_module.js:17:11
make[1]: *** [istanbul_tests] Error 1
make: *** [sdk/nodejs_all] Error 2

**Comment by EvanBoyle** [comment]
This dependency was [merged into master](https://github.com/pulumi/pulumi/pull/3424/files#diff-086221861d63aa612be766db0efc60ceR13) two days ago. 

Can you see the `callback-to-async-iterator` package in the `node_modules` folder? 
```sh
# from your pulumi/pulumi repo root
ls sdk/nodejs/node_modules/
```
If not, can you try running `make ensure` again?

**Comment by prasannapad** [comment]
I am able to get past that error by running `make ensure` first, thanks for the tip.

A few tests are still failing with the errors below. any suggestions?

31 passing (7s)
  46 failing

  1) rpc run test: empty (pwd=undefined,prog=/usr/local/src/pulumi/sdk/nodejs/bin/tests/runtime/langhost/cases/000.empty):
     Uncaught Error: spawn pulumi-language-nodejs ENOENT
      at Process.ChildProcess._handle.onexit (internal/child_process.js:264:19)
      at onErrorNT (internal/child_process.js:456:16)
      at processTicksAndRejections (internal/process/task_queues.js:80:21)

  2) rpc run test: empty.pwd (pwd=/usr/local/src/pulumi/sdk/nodejs/bin/tests/runtime/langhost/cases/000.empty,prog=./):
     Uncaught Error: spawn pulumi-language-nodejs ENOENT
      at Process.ChildProcess._handle.onexit (internal/child_process.js:264:19)
      at onErrorNT (internal/child_process.js:456:16)
      at processTicksAndRejections (internal/process/task_queues.js:80:21)

  3) rpc run test: empty.pwd.index.js (pwd=/usr/local/src/pulumi/sdk/nodejs/bin/tests/runtime/langhost/cases/000.empty,prog=./index.js):
     Uncaught Error: spawn pulumi-language-nodejs ENOENT
      at Process.ChildProcess._handle.onexit (internal/child_process.js:264:19)
      at onErrorNT (internal/child_process.js:456:16)
      at processTicksAndRejections (internal/process/task_queues.js:80:21)
Uncaught Error: spawn pulumi-language-nodejs ENOENT

**Comment by EvanBoyle** [comment]
I ran into this issue when I got started earlier this week. I believe it is related to `/opt/pulumi/` and '/opt/pulumi/bin' not being on the 'PATH'. I did a wave of updates to the [CONTRIBUTING.md](https://github.com/pulumi/pulumi/blob/master/CONTRIBUTING.md), would you mind trying to run through those steps from the beginning to see if my changes are enough to get started?

**Comment by prasannapad** [comment]
Looking better but couple of tests still failing

(timeout issue?)

375 passing (2m)
  2 failing

  1) rpc run test: provider_invokes (pwd=undefined,prog=/usr/local/src/pulumi/sdk/nodejs/bin/tests/runtime/langhost/cases/060.provider_invokes):
     Error: Timeout of 15000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
      at listOnTimeout (internal/timers.js:531:17)
      at processTimers (internal/timers.js:475:7)

  2) rpc run test: provider_in_parent_invokes (pwd=undefined,prog=/usr/local/src/pulumi/sdk/nodejs/bin/tests/runtime/langhost/cases/061.provider_in_parent_invokes):
     Error: Timeout of 15000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
      at listOnTimeout (internal/timers.js:531:17)
      at processTimers (internal/timers.js:475:7)

**Comment by EvanBoyle** [comment]
These two tests are known to fail on newer versions of node (12.11+ I believe). Let me try to find the related issue to link. In the meantime to unblock yourself you can either downgrade node, or [comment out these tests](https://github.com/pulumi/pulumi/blob/master/sdk/nodejs/tests/runtime/langhost/run.spec.ts#L901-L939).

**Comment by prasannapad** [comment]
Downgraded Node to 12.10.0 now it builds but getting 2 (different) errors  below.

BUILD:
go install -ldflags "-X github.com/pulumi/pulumi/pkg/version.Version=v1.6.0-alpha.1573752355+g5ae4149a.dirty" github.com/pulumi/pulumi
# golang.org/x/xerrors
/Users/ppadmana/go/pkg/mod/golang.org/x/xerrors@v0.0.0-20190410155217-1f06c39b4373/adaptor_go1_13.go:16:14: undefined: errors.Frame
/Users/ppadmana/go/pkg/mod/golang.org/x/xerrors@v0.0.0-20190410155217-1f06c39b4373/format_go1_13.go:12:18: undefined: errors.Formatter
make: *** [build] Error 2

**Comment by EvanBoyle** [comment]
From the error it looks like you're using go 1.13. We require v1.12 for development. Can you run `which go` and make sure that go v1.12 is on your path?

**Comment by prasannapad** [comment]
Thanks for the pointer. I'll give it a shot later today. Meanwhile, any chance we can catch these (node version, go version etc.) as _pre-flight checks_ say in one or more of the `make` targets?

**Comment by lukehoban** [comment]
> Meanwhile, any chance we can catch these (node version, go version etc.) as pre-flight checks say in one or more of the make targets?

Yes - we definitely want to improve the build scripts across the board to be clearer on required tools versions.   

We also want to document these more precisely and keep them up to date in `CONTRIBUTING.md` docs.

Currently supported versions are so far "do what CI does" which is https://github.com/pulumi/scripts/blob/master/ci/install-common-toolchain.sh, but this isn't well documented or enforced by the build.

**Comment by prasannapad** [comment]
Looks like I am almost there.. After downgrading go version to 1.12, down to a single error now. 


```
new_test.go:47: 
                Error Trace:    new_test.go:47
                Error:          Received unexpected error:
                                PULUMI_ACCESS_TOKEN must be set for login during non-interactive CLI sessions
                                github.com/pulumi/pulumi/pkg/backend/httpstate.Login
                                        /usr/local/src/pulumi/pkg/backend/httpstate/backend.go:270
                                github.com/pulumi/pulumi/cmd.currentBackend
                                        /usr/local/src/pulumi/cmd/util.go:83
                                github.com/pulumi/pulumi/cmd.runNew
                                        /usr/local/src/pulumi/cmd/new.go:115
                                github.com/pulumi/pulumi/cmd.TestCreatingStackWithArgsSpecifiedName
                                        /usr/local/src/pulumi/cmd/new_test.go:46
                                testing.tRunner
                                        /usr/local/Cellar/go@1.12/1.12.13/libexec/src/testing/testing.go:865
                                runtime.goexit
                                        /usr/local/Cellar/go@1.12/1.12.13/libexec/src/runtime/asm_amd64.s:1337
                Test:           TestCreatingStackWithArgsSpecifiedName
    new_test.go:760: 
                Error Trace:    new_test.go:760
                                                        new_test.go:49
                Error:          Received unexpected error:
                                no Pulumi.yaml project file found (searching upwards from /private/tmp/test-env253359164). If you have not created a project yet, use `pulumi new` to do so
                                github.com/pulumi/pulumi/pkg/workspace.NewFrom
                                        /usr/local/src/pulumi/pkg/workspace/workspace.go:91
                                github.com/pulumi/pulumi/pkg/workspace.New
                                        /usr/local/src/pulumi/pkg/workspace/workspace.go:71
                                github.com/pulumi/pulumi/cmd.loadStackName
                                        /usr/local/src/pulumi/cmd/new_test.go:759
                                github.com/pulumi/pulumi/cmd.TestCreatingStackWithArgsSpecifiedName
                                        /usr/local/src/pulumi/cmd/new_test.go:49
                                testing.tRunner
                                        /usr/local/Cellar/go@1.12/1.12.13/libexec/src/testing/testing.go:865
                                runtime.goexit
                                        /usr/local/Cellar/go@1.12/1.12.13/libexec/src/runtime/asm_amd64.s:1337
                Test:           TestCreatingStackWithArgsSpecifiedName
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x4ebd309]
```
  
Note: I did do `pulumi login` before starting the build and set the environment variable.
`export PULUMI_ACCESS_TOKEN = <access token from pulumi website> `

**Comment by mikhailshilkov** [comment]
It looks like the error in title has been resolved since. As it's not exactly clear what this issue is tracking, and it has been inactive for 2+ years, I'll go ahead and close it. Please report and new build errors separately.
